### ISSUE TITLE
How can we force the feeds to refresh?

### ISSUE THREAD

**Issue by mikeblakeuk** [MicrosoftDocs/azure-devops-docs#8382]
"There is a 3-6 hour delay between when a package is pushed to an upstream source and when it is available for download by your feed. "
How can we force the feeds to refresh?
It's confusing pulling a package which fails, and then a fellow developer pulling, and it works.
https://developercommunity.visualstudio.com/content/problem/1038051/how-can-we-force-the-cache-of-the-artifacts-to-ref.html

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d5f022df-36c3-34b1-cac1-ac33995a7d73
* Version Independent ID: 9634b0c8-7b14-4453-2c04-aedabbf331f1
* Content: [Upstream sources - Azure Artifacts](https://docs.microsoft.com/en-us/azure/devops/artifacts/concepts/upstream-sources?view=azure-devops#metadata-cached-from-upstream-sources)
* Content Source: [docs/artifacts/concepts/upstream-sources.md](https://github.com/MicrosoftDocs/azure-devops-docs/blob/master/docs/artifacts/concepts/upstream-sources.md)
* Product: **devops**
* Technology: **devops-artifacts**
* GitHub Login: @ramiMSFT
* Microsoft Alias: **rabououn**

**Comment by msebolt** [comment]
@mikeblakeuk Thank you for the question and with providing the product team's response.

The documentation will be updated with a brief explanation.

**Comment by mikeblakeuk** [comment]
> Hi, I have the same question.
> how can I force sync azure DevOps artifacts with the upstream source?

You shouldn't need to anymore. It is near instant.

**Comment by saostad** [comment]
for me, it took around 3 hours!

**Comment by saostad** [comment]
> > Hi, I have the same question.
> > how can I force sync azure DevOps artifacts with the upstream source?
> 
> You shouldn't need to anymore. It is near instant.

for me, it took around 3 hours!

**Comment by deMD** [comment]
> 
> 
> > Hi, I have the same question.
> > how can I force sync azure DevOps artifacts with the upstream source?
> 
> You shouldn't need to anymore. It is near instant.

And for me it's been taking a day and counting :/

**Comment by djee-ms** [comment]
I can regularly observe 3+ hours delays, it's certainly not "instant".

**Comment by danshapir** [comment]
> > Hi, I have the same question.
> > how can I force sync azure DevOps artifacts with the upstream source?
> 
> You shouldn't need to anymore. It is near instant.

There is nothing instant about that... Hours and counting.

**Comment by ramiMSFT** [comment]
Thank you for the feedbacks. Investigating

**Comment by giondo** [comment]
force refresh and see the logs?
that black box is making me loose almost 2 weeks now

**Comment by dalemyers** [comment]
This is continuing to be an issue.

**Comment by danymarques** [comment]
Same here.. Still an issue!

**Comment by danmarshall** [comment]
I have been working around this issue by deleting my `.npmrc` file, then `npm install`ing from public sources, then restoring the `.npmrc` file and committing the `package.json` and `package-lock.json` files.

**Comment by melloware** [comment]
@danmarshall Great workaround.  But this is a frustrating issue.  

I would be OK even with a "Sync" button next to the Upstream source. Especially in today's security environment where we can't wait 3-6 hours for a security patch. Hackers aren't waiting....

**Comment by danshapir** [comment]
Still an issue. This is awful!

**Comment by Der-Kraken** [comment]
Is there any workaround? Something like call an api which force the feed to look into upstream source immediately?

**Comment by ramiMSFT** [comment]
Thank you all for your feedback! I wanted to provide an update on what's happening behind the scenes. The product team is aware of this issue and they're looking into it. I will post an update once more details are available Thank you!

\+ @omokoh

**Comment by BruceHaley** [comment]
Our feed "FuseNPM" https://dev.azure.com/FuseLabs/SDK_v4/_packaging?_a=feed&feed=FuseNPM is not updating from upstream sources. The last time the api-extractor package got update was in 2018!!

**Comment by Sayan751** [comment]
Is there any update on this?

**Comment by franher** [comment]
on an npm package, it's taking hours to see the new version from the public registry on the Azure feed :(

**Comment by ramiMSFT** [comment]
> Thank you all for your feedback! I wanted to provide an update on what's happening behind the scenes. The product team is aware of this issue and they're looking into it. I will post an update once more details are available Thank you!
> 
>  +@omokoh 

Adding @jorfde
@omokoh
Any updates on this?

**Comment by Adjective-Object** [comment]
For us it requires a download attempt plus several hours of waiting for a new version of a package to be pulled in.

This is a huge pain point for me and my team, who manage packages on public npm and consume them on our internal feed.

**Comment by BruceHaley** [comment]
I pushed this package to npmjs.com on April 22: https://www.npmjs.com/package/@microsoft/orchestrator-core/v/4.15.0-dev.20220422.8d959f7h
Today is April 25, and the package is still not visible on our ADO Artifacts feed, even though the feed shows the message that it was last refreshed from upstream this morning. Our feed still shows the old package version: https://dev.azure.com/FuseLabs/SDK_v4/_artifacts/feed/SDK_JS_org/Npm/@microsoft%2Forchestrator-core/4.15.0-dev.20220218.277f659h/overview

**Comment by bodograumann** [comment]
@BruceHaley Yes, the documentation says that a package version is only visible in the web UI, after you have installed it from the mirror, iirc.
The bigger problem is, that after a new package version has been published upstream, it can take many hours until it is even possible to install that version from the mirror.

**Comment by WilliamAntonRohm** [comment]
@bodograumann -- Bodo, thank you for your follow-up.  This product-related issue is outside the realm of documentation; [please add it here](https://developercommunity.visualstudio.com/search?space=21).

**Comment by kyle-rader** [comment]
I'm still seeing multi-day lag from when I first try to install a newly published version and when it can actually be installed. 
I'm trying https://gist.github.com/mikeblakeuk/7436c4861e363b135a492ea34c74eb66 but I'm getting "feed does not exist" :( anyone know if that endpoint has changed?

**Comment by AbstractionsAs** [comment]
I'm experiencing this well

**Comment by carlin-q-scott** [comment]
@kyle-rader 

> I'm still seeing multi-day lag from when I first try to install a newly published version and when it can actually be installed. I'm trying to no avail.

What I've found is that the user performing the request for the package must have owner/admin permissions in order for the request to pull in an upstream package (version). This same limitation applies to the API request you shared, as well as `nuget install $packageName` or `dotnet add package $packageName`. 

From [looking at this permissions table](https://docs.microsoft.com/en-us/azure/devops/artifacts/feeds/feed-permissions?view=azure-devops#permissions-table), I'd imagine the permission needed would be: "Save packages from upstream sources". However, my ServiceToken has that permission and the call fails for it, even though it succeeds using my Owner level Personal Access Token (PAT).

**Comment by carlin-q-scott** [comment]
@WilliamAntonRohm 
> This product-related issue is outside the realm of documentation.

I think the documentation should call out this limitation of the upstream feeds feature as it makes for a poor user experience and still is an issue two years later. I think there should be a published workaround in the documentation until this gets fixed. And the workaround seems to be a permission issue that doesn't allow for an automated workaround without sharing my PAT with the build system.

**Comment by carlin-q-scott** [comment]
I reported [this issue as instructed to the developer community](https://developercommunity.visualstudio.com/t/private-NuGet-feed-not-providing-upstrea/10067650?space=21&q=nuget+upstream+feed), but as you can see in my issue, it has been reported and ignored thousands of times. But maybe they'll fix it this time now that they know it's a permissions issue.

**Comment by WilliamAntonRohm** [comment]
@carlin-q-scott -- Carlin, thank you for your considered follow-up.  I upvoted your issue in the Developer Community.

@ramiMSFT -- Rami, is there a way to raise this issue further?

**Comment by ramiMSFT** [comment]
@omokoh @jorfde This issue is still pending. Can we take a look at this

**Comment by DavidVollmers** [comment]
Also a blocker for me. This is getting frustrating...

**Comment by alaingiller** [comment]
The question was "How can we force the feeds to refresh?" and it seams this problem still unresolved. 
If a public package get a new release in npm (let says a emergency security fix) we still have to wait 6h before we can use it.
For me this bug should not be closed.

**Comment by tomastrajan** [comment]
This is not resolved as for most people in the issue the described "solution"

> It may take up to 6 hours to sync up. A delay of 3-6 hours can be expected in some cases but usually the sync up happens faster.

is the problem itself, we need to get the new version (of critical bugfix for example) faster or at least be able to force refresh for some packages explicitly while default behavior staying 3-6 hours...

**Comment by arefinsami** [comment]
I found that packages in my ADO Feed never updates! Now I want to get a new version of a package due to security issue and the feed is not fetching the new version. Frustrating.

**Comment by rvdginste** [comment]
@ramiMSFT 

> We have two updates that we will be deploying soon that will have a direct impact on this issue and will greatly improve the overall user experience.

Can you tell us more about these updates, what their impact is and when they will be rolled out?

> Packages from public registries are synced using periodic refresh jobs. It may take up to 6 hours to sync up. A delay of 3-6 hours can be expected in some cases but usually the sync up happens faster.

This was known 2 years ago and is the reason that the issue was created. Is there a mechanism to force a refresh for a specific package and version?

This is blocking our team and making Azure DevOps feeds unworkable.  We maintain public packages that we use in private projects. Every time we create a new version of those public packages, we have to wait hours before the new version is available in the private feed.

**Comment by rafaelmoni** [comment]
I also face this error where I work, and the workaround we found for this is to run the pipeline once with our `.npmrc` **registry** pointing to **npmjs registry URL**, then rollback and run again with our **private registry**, if you have private package (as we do), you need to remove it from your **package.json**, just make sure your install succeed with npmjs registry first, this will force azure to sync their cache.

**Comment by gavin-theta** [comment]
We have the same issue - stale packages in a feed that is configured with upstream feeds. How do we refresh the packages and pull in the latest?

**Comment by aspeckt112** [comment]
I'm not trying to be rude or anything, but this issue is over two years old. How do we know have a manual sync button yet? Seriously, what's the hold up?

I LOVE the idea of upstream sources, version locking and tightly controlling this stuff. I lead a small team, and we're very keen on the idea of controlling what versions we use across our product line up.

**Comment by ahmadyo** [comment]
> The question was "How can we force the feeds to refresh?" and it seams this problem still unresolved. If a public package get a new release in npm (let says a emergency security fix) we still have to wait 6h before we can use it. For me this bug should not be closed.

It occurs not only form npm.js as external source, also for Nuget packages from another azure feed defined as upstream feed.

**Comment by cmeeren** [comment]
I to find the current behavior problematic. I want to fix a bug in our systems that requires a fix in a public NuGet package that I maintain. I fix that and publish it to NuGet, which is near instantaneous. Then I have to wait for 3-6 hours before actually publishing the fix to our systems. This is not a good experience, and I am close to adding the public NuGet feed as a secondary feed (with the caveats that entails).

If sync can't be instantaneous, we need an escape hatch where we can go a package on Azure Artifacts and manually check that package for updates.

**Comment by Chapi15** [comment]
I found a workaround, however I'm not using **npm**; in my case it's with **PyPi** the issue. 

These are the steps I followed:

1. Delete package from your system or dev environment
2. Delete package from your Azure DevOps feed
3. Install the package via CLI **including the URL of my feed** inside Azure DevOps
    `pip install -U --extra-index-url https://pkgs.dev.azure.com/[Organization]/_packaging/[feed_name]/pypi/simple/ package_name` 

My theory is that your installer (npm, pip, etc) looks for the package in the upstream source if it's not found in your Azure DevOps feed, by doing that the package's version is updated.

I suppose @carlin-q-scott was right regarding the permissions of your feed, you need to be admin for pulling a newer version of the package:

> What I've found is that the user performing the request for the package must have owner/admin permissions in order for the request to pull in an upstream package (version). This same limitation applies to the API request you shared, as well as `nuget install $packageName` or `dotnet add package $packageName`.
> 
> From [looking at this permissions table](https://docs.microsoft.com/en-us/azure/devops/artifacts/feeds/feed-permissions?view=azure-devops#permissions-table), I'd imagine the permission needed would be: "Save packages from upstream sources". However, my pipeline System.AccessToken has that permission and the call fails for it, even though it succeeds using my Owner level Personal Access Token (PAT).

#8382

**Comment by ViRuSTriNiTy** [comment]
Today is, again, a day where the sync between upstream and downstream feed is not working properly or is delayed for a substantial amount of time. It worked like a charm for months now on my end but today not. So it must be in the background that is causing this. Either introduce a sync button or other option to force the sync. Reconfiguring my setup all the time when its occasionally not working is tedious to say the least.

**Comment by ViRuSTriNiTy** [comment]
@gavin-theta 

> We have the same issue - stale packages in a feed that is configured with upstream feeds. How do we refresh the packages and pull in the latest?

Since this issue occasionally occurs out of nowhere, I just recommend pushing the packages to the downstream feed. In my case I added an optional parameter to my publish script to accept a specific feed. When the issue occurs, I specify the downstream feed otherwise the default is used, which is the upstream feed.

**Comment by cmeeren** [comment]
> Today is, again, a day where the sync between upstream and downstream feed is not working properly or is delayed for a substantial amount of time.

For me it's like that more or less all of the time. (I suspect that the few times there's just a short delay, it's because some scheduled sync happened to occur just after publishing to the upstream feed.)

> Since this issue occasionally occurs out of nowhere, I just recommend pushing the packages to the downstream feed.

Can't easily be done for OSS projects. I don't want to "pollute" the OSS libraries/frameworks I maintain by making them authenticate against and push to another feed of my choosing, in addition to the public NuGet feed. And maintainers of other OSS packages will certainly not do that.

**Comment by mikeblakeuk** [comment]
@ramiMSFT please reopen this

**Comment by chaehonglee** [comment]
Hi, I need this addressed as well. None of the suggested solutions, deleting the package, etc. work with reliability. Sometimes it instantly syncs with the newest version, sometimes I have to wait couple days.

**Comment by ramiMSFT** [comment]
My apologies for the delayed response, and thank you all for your patience and your great feedback. We recently rolled out two updates that will help streamline the process of ingesting packages from upstreams. Now that the features are public, I can share more details and how this is going to help this sync up scenario. 

> Can you tell us more about these updates, what their impact is and when they will be rolled out?

The first update is to address the last sync up from upstreams. Please read about it here:  [Health status](https://learn.microsoft.com/en-us/azure/devops/artifacts/concepts/upstream-sources?view=azure-devops#health-status). And then we added the [Search for packages in upstream sources](https://learn.microsoft.com/en-us/azure/devops/artifacts/how-to/search-upstream?view=azure-devops) feature to allow users to find and ingest packages right from their feed's UI.

> Is there a mechanism to force a refresh for a specific package and version?

No we don't have a mechanism to perform a force refresh. This is by design because we cannot allow feed readers to change the content of the feed (by ingesting packages from upstreams).

We suggest using the **search in upstreams** feature to search and ingest packages directly from the UI and verify if your saved package is there (in this feed). 
![image](https://user-images.githubusercontent.com/60942540/196820806-a84b8fe2-7417-4743-8b59-8a485fc6d294.png)

**Comment by cmeeren** [comment]
> We suggest using the **search in upstreams** feature to search and ingest packages directly from the UI and verify if your saved package is there (in this feed).

I'm not sure what you mean. I just published a package to nuget.org and tried this search. It does not seem to have ingested anything; I still do not see the new version in the Azure Artifacts UI.



> No we don't have a mechanism to perform a force refresh. This is by design because we cannot allow feed readers to change the content of the feed (by ingesting packages from upstreams).

I'm not sure what you mean by "feed readers" or "ingest". But I am signed in to Azure DevOps with full access to everything. Why shouldn't I be able to do something so that I can consume the updated package from my Azure Artifacts feed immediately instead of having to wait for 3-6 hours?

**Comment by ViRuSTriNiTy** [comment]
@ramiMSFT Thanks for implementing an option to manually force the sync of an upstream package, much appreciated.

But still, there seems to be an issue, this time on the dev box. When I run `dotnet add package ...` it often does not work because the manually synced package cannot be found. I think in this case the caching is the culprit as I can see the following line in console:

`info :   CACHE https://pkgs.dev.azure.com/.../index.json`

How can I disable the use of a cache in this case? Where can I find the cache on the dev box?

**Comment by cmeeren** [comment]
> @ramiMSFT Thanks for implementing an option to manually force the sync of an upstream package, much appreciated.

How did you get this to work? As mentioned in my previous comment, I just published a package to nuget.org and tried the "upstream package search" indicated by @ramiMSFT. It immediately put me on the Azure Artifacts page for that package, and does not seem to have ingested anything; I still do not see the new version in the Azure Artifacts UI - only the previous version.

**Comment by ViRuSTriNiTy** [comment]
@cmeeren .... not sure about the scenario you are having difficulties with. I only use the Azure DevOps feeds to publish my packages whereas feed B is the upstream of feed A. Before @ramiMSFT posted the new feature I often had to wait a substantial amount of time until feed A recognized the new package in feed B. Now I can directly go to feed A, select the package, then select `Upstream versions`, then click on the 3-dot menu on the right side of the new package version, then select `Save to feed`. This works in my case ... when there wouldn't be the cache issue as described.

**Comment by cmeeren** [comment]
Ah, you're thinking about this?

<img width="666" alt="image" src="https://user-images.githubusercontent.com/7766733/197805603-73b3c9ab-6132-441b-b92a-67e36c9f5483.png">

I can click "Save to feed" here, but my issue is that it takes a long time before the newly published upstream version is visible in this UI in the first place.

**Comment by cmeeren** [comment]
> Ok, I don't have this delay, most likely because it is all Azure DevOps feeds on my side.

Yes, this delay happens for public feeds like nuget.org. Remedying this delay is what this issue was created for. 🙂

**Comment by ViRuSTriNiTy** [comment]
@cmeeren Now I am also experiencing the delay but I can see that I am now allowed to push the packages to the require feed directly without removing the upstream source configuration. It seems like MS is working on the feed sync in the background as currently each day brings a new surprise.

**Comment by drvup** [comment]
Thanks @ramiMSFT for your feedback and the introduction to 'Search in upstream feeds'. It's a great idea, but it is not working as it should be. 

Having multiple upstreams, all hosted on AzureDevOps, I thought I can search for (npm) packages in 1 artifact-stream on the other, if they are related to each other via upstreams. But this isn't working and does not deliver any result. I am searching with right name of package, with and without versions, it's not working. 

Please keep on investigating and find a way to us AzureDevOps users, who are frustrated about not having a simple sync button :/ 

Thanks.

**Comment by mikeblakeuk** [comment]
This should be re-opened.

Each time Microsoft release a .net patch (e.g. 6.0.10 to 6.0.11) release our builds fail with
```
    D:\a\1\s\Project.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Runtime.win-x64 with version (= 6.0.11)
       D:\a\1\s\Project.csproj : error NU1102:   - Found 29 version(s) in InternalFeed [ Nearest version: 6.0.10 ]
```
We have a Product that restores from **InternalFeed** that has an upstream of **CentralFeed** that pulls from NuGet.

The InternalFeed, can't pull from NuGet.

A work around is to restore from CentralFeed first. that pull from Microsoft. Then pull from InternalFeed

**Comment by mjbvz** [comment]
@mikeblakeuk I don't work on dev ops but have hit this issue multiple times while working on VS Code

Having to wait multiple hours for a feed to refresh is impacting our workflow and has broken our CI multiple times now. I filed an issue on dev-ops developer community but unfortunately they do not seem to consider this a bug: https://developercommunity.visualstudio.com/t/Feed-update-for-newly-published-public-n/10182641

**Comment by thomasnal** [comment]
@cmeeren Can this be caused by configuration of the feed? Azure [docs](https://learn.microsoft.com/en-us/azure/devops/artifacts/concepts/upstream-behavior?view=azure-devops&tabs=nuget%2Cget) describe situations when the feed blocks download from the public registry. When there is already a version of the package in the feed it blocks download of new version from the public registry, depending on some configuration.

There is the setting `Allow externally-sourced versions` that can be turned on to remove the block.  I've not seen the thread saying if you have this set on.

Note: I came to report another upstream issue with Python feeds and checking other opened issues.

**Comment by cmeeren** [comment]
That's a different issue.

**Comment by svdHero** [comment]
> Ah, you're thinking about this?
> 
> <img alt="image" width="666" src="https://user-images.githubusercontent.com/7766733/197805603-73b3c9ab-6132-441b-b92a-67e36c9f5483.png">
> 
> I can click "Save to feed" here, but my issue is that it takes a long time before the newly published upstream version is visible in this UI in the first place.

I have exactly the same problem with NuGet packages, although both feeds are from private projects on Azure DevOps. No public upstream whatsoever.
Project A is upstream of project B. When I build a new version of A, said list in feed B does not show the new version. Hence, there is no "Save to feed" button to klick.

![image](https://user-images.githubusercontent.com/24250824/211340224-44454d91-c402-4ab7-8ec9-2f38b19b9d7e.png)

![image](https://user-images.githubusercontent.com/24250824/211340299-b317b481-54aa-426b-b2f7-9786c22c521a.png)


@ramiMSFT Can this be re-opened, please? It is obviously not fixed, yet.

**Comment by Ptipoi-jf** [comment]
I have exactly the same problem as mentioned by svdHero. 

I have two private feeds (A & B), A is upstream of B. 

Today, I built a new version of my nuget package in feed A, and it was immediately seen on B. 
BUT I need to upload a new version 30m later (bug fix) in feed A, and then the bug fix is not synced on my feed B even after waiting for a few hours.

feed A :
![image](https://user-images.githubusercontent.com/9008010/211798379-82bc724a-6f8c-46d6-ba40-7ac813ce786a.png)

feed B :
![image](https://user-images.githubusercontent.com/9008010/211798629-b744290c-7ab8-4d18-bb8c-5ca250a0c423.png)

`dotnet restore` on the feed B did not work to get the latest version from feed A

**Comment by DavidVollmers** [comment]
Why is this always closed when nothing changes?!

Waiting 3-6 hours for a feed sync to happen is just not an acceptable solution for any one working on SaaS products.
We need either much higher throughput or a way to force a sync for a certain package as a Feed Owner.

I don't get why this is ignored...

**Comment by svdHero** [comment]
I don't get it either. Especially since in my case the feed is never synced and the manual workaround doesn't work either. This renders ADO absolutely useless for us when it comes to saving artifacts. It's a shame that we have to look elsewhere for this.

@ramiMSFT could you please reopen this issue and look into this again?

**Comment by DavidVollmers** [comment]
I created a new issue here: https://developercommunity.visualstudio.com/t/Waiting-3-6-hours-for-a-feed-sync-to-hap/10272792?q=How+can+we+force+the+feeds+to+refresh%3F

**Comment by Moerup** [comment]
Having the exact same problem as described above and it's blocking us from using Artifacts feed correctly, this is not fixed, please reopen.

**Comment by DavidVollmers** [comment]
> “Is there a mechanism to force a refresh for a specific package and version?” , from our escalated resource that the answer is ‘No’ and we don’t have a mechanism to perform a force refresh. **This is by design because we cannot allow feed readers to change the content of the feed (by ingesting packages from upstreams).**

I created a feature request for a solution on this issue which I invite everyone here to upvote:
https://developercommunity.visualstudio.com/t/Option-to-refresh-feed-sync/10279501

**Comment by yurim108** [comment]
Same problem with two DevOps projects. I am using UPack to pass several artifacts from an artifact feed of one project into an artifact feed of another project. Sometimes the new version of one artifact can be found "at once", but few seconds later searching for the new version of another artifact does not bring the expected result.

**Comment by yurim108** [comment]
@WilliamAntonRohm, @ramiMSFT the issue is still there and is quite disturbing. Why ist the ticket closed? Have you provided a solution? Or are you not going to fix that? Please state it clearly.

**Comment by DavidVollmers** [comment]
@yurim108 This is not an issue as stated by Microsoft. (see my last comment on this thread)

You can and should however upvote my feature request:
https://developercommunity.visualstudio.com/t/Option-to-refresh-feed-sync/10279501

**Comment by kyle-rader-msft** [comment]
Yet another multi-day delay in a feature because we can't get a new package version :(

**Comment by ChristophHornung** [comment]
I ran into the same problem and found out the correct HEAD request updates the nuget package cache in the azure artifact feed.

I wrapped that into a small project [here](https://github.com/ChristophHornung/AzureDevOpsNugetRefresher/releases/tag/v0.0.1), it seems to work for me, if someone else can test it I am happy about any feedback.

**Comment by kyle-rader-msft** [comment]
Thanks @ChristophHornung , I will try that approach. Here am I again, my semi-annual multi-day feature blockage because I can't build in CI or PR pipelines due to packages not being found. This time trying to use the crates.io upstream for Cargo packages :(.

**Comment by johnterickson** [comment]
There's a few different things going on in this thread.
1. There's a limitation in the system that when a new package version is uploaded to a public source (e.g. npmjs, pypi, crates.io) then there can be a few hours of delay - there should never be days.  Upstreams from one AzDO feed to another should be on the order of seconds.
2. You might not see a new version of a package if you have mixed public and private sources for package https://learn.microsoft.com/en-us/azure/devops/artifacts/concepts/upstream-behavior?view=azure-devops&tabs=nuget%2Cget
3. You will not be able to ingest new versions into a feed unless you have `Collaborator` or higher access. 
4. Many clients have *local* caches so you will not see a new package version even if Artifacts has it.  e.g. `http-cache` https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-nuget-locals

@kyle-rader-msft hit #5 a bug in how the service handles "packages that have a particular property that used to be allowed on crates.io and no longer are".  This bug is being fixed.

**Comment by DavidVollmers** [comment]
@johnterickson In todays world it is not acceptable to have a "few hours" of delay when needing to push hotfixes or nightly deployments.

This literally implies to not use Azure DevOps private feeds or only use them...

**Comment by johnterickson** [comment]
To make sure we're on the same page, there is no delay when retrieving package versions that have already existed in the public source. If you find that you suddenly need to upgrade the version of a package (and that version has existed in the public source for more than a few hours) then there is no delay in consuming it.

This does not cover all scenarios, but I realized my earlier wording was ambiguous.

**Comment by DavidVollmers** [comment]
@johnterickson In which scenario would I need to upgrade the version of a package which is already uploaded to the public registry for more than a few hours?

The problem is, for example, when you push a new package version (to an existing package) to the public npm registry it will sometimes take 3-6 hours to be fetched by the private ADO registry.

The default use case is to have a private registry which mirrors the public registry so that you can consume both public and private packages. And in case you have a private package/application which depends on a public package you are literally f*cked when trying to hotfix something. 

I really don't know why this is played down so much by microsoft...

**Comment by kyle-rader** [comment]
To follow up, I hit the combo of a 7 hour wait on many packages, and then
the one particular issue John mentioned for 1 specific crate that had a "-"
in the name, but was cached as not existing with an "_". And our ADO team
was very prompt in fixing that particular package on our internal feed.

**Comment by StefanKoell** [comment]
Guys, this is just a joke! Waiting for hours (or even days) to get an updated version showing up in the DevOps package feed is unacceptable. I'm happy to report that we are in the process of moving away from Azure DevOps because of this and a lot of other issues where MS just doesn't want to listen to its customers.

**Comment by patrickcdsouza** [comment]
NuGet packages pushed to an upstream source are not seen on a feed using that upstream source instantly. I know this is an old issue, but has there been any workaround for this?

 I read a few articles that said there is a three-hour period before the refresh happens. Can this be lowered to a smaller period?

**Comment by aferlim** [comment]
same problem. Plz fix it

**Comment by WKnak** [comment]
For me the problem was the format of the package version. 

I was adding the word "-release" after the version, but in fact, any dash/string after the version is considered a non-stable version, doesn't matter if it is "release". 😅

So, my package version "1.0.14-release" was only showing up in the packages lists of Visual Studio if I check "include prerelease", even though I had promoted it to Release channel in the Artifacts Feed.

Then I edit the pipeline and removed the "release" word, and the generated version "1.0.15" becomes visible as stable as soon as I promote it to Release on the Feed. 😍
